src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java

```diff
@@ -1059,14 +1060,28 @@ void setPackageOrDeclaringType(CtTypeReference<?> ref, CtReference declaring) {
 				javaLangPackageReference.setSimpleName("java.lang");
 				ref.setPackage(javaLangPackageReference);
 			} catch (NoClassDefFoundError | ClassNotFoundException e) {
-				// in that case we consider the package should be the same as the current one. Fix #1293
-				ref.setPackage(jdtTreeBuilder.getContextBuilder().compilationUnitSpoon.getDeclaredPackage().getReference());
+				ContextBuilder ctx = jdtTreeBuilder.getContextBuilder();
+				boolean isNoClasspath = jdtTreeBuilder.getFactory().getEnvironment().getNoClasspath();
+				if (isNoClasspath && containsStarImport(ctx.compilationunitdeclaration.imports)) {
+					// If there is an unresolved star import in noclasspath,
+					// we can't tell which package the type belongs to (#3337)
+					CtPackageReference pkgRef = jdtTreeBuilder.getFactory().Core().createPackageReference();
+					pkgRef.setImplicit(true);
+					ref.setPackage(pkgRef);
+				} else {
+					// otherwise the type must belong to the CU's package (#1293)
+					ref.setPackage(ctx.compilationUnitSpoon.getDeclaredPackage().getReference());
+				}
 			}
 		} else {
 			throw new AssertionError("unexpected declaring type: " + declaring.getClass() + " of " + declaring);
 		}
 	}
 
+	private static boolean containsStarImport(ImportReference[] imports) {
+	    return imports != null && Arrays.stream(imports).anyMatch(imp -> imp.toString().endsWith("*"));
```

Comment:
This will return true if there's _any_ star import, which is too restrictive. For example, if there's a `java.util.*`, then this will return true and we won't assign the CU's package to the type. But if the star import is actually available, which should be the case for stdlib packages, depending on JDK version, then it should be fine to guess assign the CU's package to the type.

Is there any way to check which imported packages are actually available at this stage?

Reply:
> Is there any way to check which imported packages are actually available at this stage?

Do you mean which subpackages under `.**`are available in the classpath?

Reply:
No, I was rather thinking of just checking if the containing package is available. For example, if any type from `java.util` is available on the classpath, then we conclude that all types from `java.util.*` are available on the classpath. If that is the case, we assume that had `SomeClass` actually belonged to `java.util`, it would too have been on the classpath too, and since it is not we assign it to the compilation unit's package.

But as I typed the above paragraph I realized a few things. First, third party packages are often split across projects leading to packages possibly being partially available, so the assumption "all types from the `x.y.*` import are available if any type from package `x.y` is available" is not necessarily true (think e.g. `org.apache`). Second, differing versions of the standard library may cause issues. Third, accounting for what I just pointed out makes this incredibly complex.

I think the simple and probably sufficient solution is to simply not guess the package if the type is not on the classpath, and there is any `*` import.